src/lib/formats/cp68_dsk.cpp

```diff
@@ -0,0 +1,266 @@
+// license:BSD-3-Clause copyright-holders:Barry Rodewald, 68bit
+/*
+ * cp68_dsk.c  -  CP68 compatible disk images
+ *
+ *  Created on: 02/08/2022
+ *
+ * This CP68 floppy disk image support leverages the wd177x_format support with
+ * the few differences handled by format variations.
+ *
+ * The CP68 Disk format is as follows:
+ *
+ * Geometry:: 18 Sectors * 35 Tracks * 128 Bytes * 1 Side
+ *
+ * Track 0:: Boot Sectors 0, 1, 2; Directory Sectors 4-18
+ * Tracks 1-34::  Sectors 1-18
+ *
+ * System Information Record:: Sector 2, Bytes $7A-$7F
+ *         $7A - System Link Start Track
+ *         $7B - System Link Start Sector
+ *         $7C - System Link End Track
+ *         $7D - System Link End Sector
+ *         $7E - Free Chain Start Track
+ *         $7F - Free Chain Start Sector
+ *
+ * The formats below include interleaved sectors to improve performance. The
+ * interleave has been chosen to perform well on slower hardware and software
+ * combinations while still offering some performance improvement. Tighter
+ * interleaving may well be possible but it depends on the systems and
+ * software.
+ *
+ * Note about Gap Bytes::
+ *         The gap bytes used here for MAME are not standard.  The reason is
+ *         with 4,000 time quanta for each cell only allows for 50,000 Cells
+ *         total on the disk.  The original Gap Bytes as shown in the manual
+ *         https://archive.org/details/CP-68_An_M6800_Operating_System_1979/page/80/mode/2up?q=GAP
+ *         wind up with too many cells to fit on the disk. This has been
+ *         adjusted to the minimum allowable according to the WD1771 data
+ *         sheet of 8, 11, 11
+ *
+ * Note: This code is compatible with the Disk images from deramp.com (Mike
+ * Douglas) at the following location.  It is _NOT_ compatible with the SimH
+ * disk images:
+ *
+ * https://deramp.com/downloads/swtpc/software/CP68/
+ */
+
+#include "cp68_dsk.h"
+
+#include "imageutl.h"
+
+#include "ioprocs.h"
+
+
+cp68_format::cp68_format() : wd177x_format(formats)
+{
+}
+
+const char *cp68_format::name() const
+{
+	return "cp68";
+}
+
+const char *cp68_format::description() const
+{
+	return "CP68 compatible disk image";
+}
+
+const char *cp68_format::extensions() const
+{
+	return "dsk";
+}
+
+int cp68_format::identify(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants) const
+{
+	int type = find_size(io, form_factor, variants);
+
+	if (type != -1)
+		return FIFID_SIZE;
+	return 0;
+}
+
+
+int cp68_format::find_size(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants) const
+{
+	uint64_t size;
+	if (io.length(size))
+		return -1;
+
+	uint8_t boot0[128];
+	sysinfo_sector_cp68 info;
+	size_t actual;
+
+	// Look at the boot sector.
+	io.read_at(128 * 0, &boot0, sizeof(boot0), actual);
```

Comment:
Given this is new code, can you please add error checking?  This can return an error result, and the read can be truncated if it hits EOF.  Same for the other reads.